e2e/e2e_test.go

```diff
@@ -81,23 +97,253 @@ var _ = Describe("Whereabouts functionality", func() {
+
+				k8sIPAM, err = kubeClient.NewKubernetesIPAMWithNamespace("", types.IPAMConfig{
+					Kubernetes: types.KubernetesConfig{
+						KubeConfigPath: envVars.kubeconfigPath,
+					},
+				}, testNamespace)
+				Expect(err).NotTo(HaveOccurred())
+
+				replicaSet, err = clientInfo.provisionReplicaSet(
+					replicaCount,
+					podTierLabel(rsName),
+					podNetworkSelectionElements(testNetworkName),
+				)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				By("deleting replicaset with whereabouts net-attach-def")
+				Expect(clientInfo.deleteReplicaSet(replicaSet)).To(Succeed())
+			})
+
+			It("allocates each IP pool entry with a unique pod IP", func() {
+				const (
+					maxPodsPerNode  = 110
+					rsSteadyTimeout = 2 * rsCreateTimeout
+				)
+
+				By("creating max number of pods and checking IP Pool validity")
+				for i := 0; i < envVars.numThrashIter; i++ {
+					// Set to 0 replicas
+					replicaCount = int32(0)
+					replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
+						replicaCount,
+						podTierLabel(rsName),
+						podNetworkSelectionElements(testNetworkName),
+					))
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, label, replicaSet, rsSteadyTimeout)).To(Succeed())
+
+					// Check for 0 IP pool allocations
+					ipPool, err := k8sIPAM.GetIPPool(context.Background(), ipPoolName)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ipPool.Allocations()).To(BeEmpty())
+
+					// Increase to desired number of replicas for test
+					allPods, err := clientInfo.Client.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					replicaCount = int32((envVars.numComputeNodes*maxPodsPerNode - (len(allPods.Items))) * envVars.fillPercentCapacity / 100)
+					replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
+						replicaCount,
+						podTierLabel(rsName),
+						podNetworkSelectionElements(testNetworkName),
+					))
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, label, replicaSet, rsSteadyTimeout)).To(Succeed())
+
+					Expect(iPPoolConsistency(envVars.kubeconfigPath, k8sClient, k8sIPAM, clientInfo.Client)).To(Succeed())
+				}
+
+				By("removing replicas and expecting 0 IP pool allocations")
+				// Set to 0 replicas
+				replicaCount = int32(0)
+				replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
+					replicaCount,
+					podTierLabel(rsName),
+					podNetworkSelectionElements(testNetworkName),
+				))
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, label, replicaSet, rsSteadyTimeout)).To(Succeed())
+
+				// Check for 0 IP pool allocations
+				ipPool, err := k8sIPAM.GetIPPool(context.Background(), ipPoolName)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(ipPool.Allocations()).To(BeEmpty())
```

Comment:
this clearly belongs on the test teardown (after each section).

I also don't understand why you scale down the replica set. Shouldn't you just remove it ?... Without it, the replicaset leaks - but has no pods.

Furthermore, check if you can encapsulate this into a neat function.

Reply:
Moved to `AfterEach`. :+1:  

I can't just remove the replicaset and do nothing else however. After the test finishes, the replicas may still be in `Terminating` state. This will effect Martin's test now, and maybe in the future someone else's tests. While this does add a little bit of code, it guarantees 0 IP pool allocations after the test runs.

I put all instances of this kind of code in a neat function like you suggested, as well.

Reply:
Yes, that is correct: you should delete the replica set **and** wait for the pods to be gone.